src/cargo/ops/cargo_install.rs

```diff
@@ -114,14 +107,90 @@ pub fn install(root: Option<&str>,
         try!(fs::remove_dir_all(&target_dir));
     }
 
+    let mut installed_names = HashSet::new();
+    let mut installed = Vec::new();
+
+    // Move the temporary copies into `dst`.
+    let install_result = {
+        let mut try_install = || -> CargoResult<()> {
+            for bin in compile.binaries.iter() {
+                let bin = bin.file_name().unwrap();
+                let src = staging_dir.path().join(bin);
+                let dst = dst.join(bin);
+                try!(config.shell().status("Installing", dst.display()));
+                try!(fs::rename(&src, &dst).chain_error(|| {
+                    human(format!("failed to move `{}` to `{}`", src.display(),
+                                  dst.display()))
+                }));
+                installed_names.insert(bin.to_string_lossy().into_owned());
+                installed.push(dst);
+            }
+            Ok(())
+        };
+        try_install()
+    };
+    // If there was partial success postpone examination of the result until
+    // after updating the metadata.
+    if !installed.is_empty() {
+        try!(install_result);
+    }
+
+    let mut to_remove = Vec::new();
+
+    if force {
+        let duplicate_pkgs = duplicates.values()
+                                       .filter_map(|v| v.clone())
+                                       .collect::<HashSet<_>>();
+        // Remove stale records of overwritten binaries.
+        for bin in installed_names.iter() {
+            if let Some(&Some(ref p)) = duplicates.get(bin) {
+                if let Some(set) = list.v1.get_mut(p) {
+                    set.remove(bin);
+                }
+            }
+        }
+        // If `--bin` or `--example` weren't specified, uninstall obsolete
```

Comment:
All this time I'd been assuming it's a good idea for `cargo install -f foo` to completely remove the remnants of the old package by default because that's probably what the user wants. :) If that's not the case I should probably add a hint "run 'cargo uninstall foo:{previous version}' to remove remaining binaries."